### PR TITLE
Cancel reconciliation when the guest cluster is not yet available

### PR DESCRIPTION
**Pull request by rossf7**
Towards giantswarm/giantswarm#1902

Cancels reconciliation in the namespace and chartconfig resources if the guest cluster is not yet up.

Tested on ginger.

```
{"caller":"github.com/giantswarm/cluster-operator/pkg/v3/guestcluster/guestcluster.go:109","function":"GetCurrentState","level":"debug","message":"created K8s client for the guest cluster","object":"/apis/core.giantswarm.io/v1alpha1/namespaces/default/awsclusterconfigs/09i83-aws-cluster-config","resource":"namespacev3","time":"2018-05-22T17:31:33.333547+00:00"}
{"caller":"github.com/giantswarm/cluster-operator/pkg/v3/resource/namespace/current.go:44","function":"GetCurrentState","level":"debug","message":"looking for the namespace in the guest cluster","object":"/apis/core.giantswarm.io/v1alpha1/namespaces/default/awsclusterconfigs/09i83-aws-cluster-config","resource":"namespacev3","time":"2018-05-22T17:31:33.333584+00:00"}
{"caller":"github.com/giantswarm/cluster-operator/pkg/v3/resource/namespace/current.go:54","function":"GetCurrentState","level":"debug","message":"guest cluster is not available","object":"/apis/core.giantswarm.io/v1alpha1/namespaces/default/awsclusterconfigs/09i83-aws-cluster-config","resource":"namespacev3","time":"2018-05-22T17:31:33.37102+00:00"}
{"caller":"github.com/giantswarm/cluster-operator/pkg/v3/resource/namespace/current.go:59","function":"GetCurrentState","level":"debug","message":"canceling resource reconciliation for custom object","object":"/apis/core.giantswarm.io/v1alpha1/namespaces/default/awsclusterconfigs/09i83-aws-cluster-config","resource":"namespacev3","time":"2018-05-22T17:31:33.371056+00:00"}
```

### CODE REVIEW

File: pkg/v3/guestcluster/error.go [hunk 1]
```diff
@@ -1,9 +1,63 @@
...
+			return true
+		}
+	}
+
+	if c == guestAPINotAvailableError {
+		return true
+	}
+
+	return false
+}
```
Comment (rossf7) [inline pkg/v3/guestcluster/error.go]: This logic is taken from @fgimenez 's changes in `giantswarm/helmclient`. It felt wrong to import this in resources that don't use Helm. If anyone has good ideas for structuring the libraries here they are very welcome!

File: pkg/v3/resource/chartconfig/current.go [hunk 2]
```diff
@@ -46,6 +46,16 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
 
 		return nil, nil
 
+	} else if guestcluster.IsGuestAPINotAvailable(err) {
```
Comment (fgimenez) [inline pkg/v3/resource/chartconfig/current.go]: Good catch adding it to chartconfig too :+1:

File: pkg/v3/guestcluster/error.go (hunk 1 above)
Comment (kopiczko) [reply pkg/v3/guestcluster/error.go]: Maybe we should have `github.com/giantswarm/errors`?

Comment (fgimenez) [reply pkg/v3/guestcluster/error.go]: :+1: a library makes sense to me.

**Pull request merged**